saleor/graphql/payment/mutations.py

```diff
@@ -1229,3 +1223,153 @@ def perform_mutation(cls, root, info, **data):
             code = error_enum.MISSING_TRANSACTION_ACTION_REQUEST_WEBHOOK.value
             raise ValidationError(str(e), code=code)
         return TransactionRequestAction(transaction=transaction)
+
+
+class TransactionEventReport(ModelMutation):
+    already_processed = graphene.Boolean(
+        description="Defines if the reported event hasn't been processed earlier."
+    )
+    transaction = graphene.Field(
+        TransactionItem, description="The transaction related to the reported event."
+    )
+    transaction_event = graphene.Field(
+        TransactionEvent,
+        description=(
+            "The event assigned to this report. if `alreadyProcessed` is set to `true`,"
+            " the previously processed event will be returned."
+        ),
+    )
+
+    class Arguments:
+        id = graphene.ID(
+            description="The ID of the transaction.",
+            required=True,
+        )
+        psp_reference = graphene.String(
+            description=("PSP Reference of the event to report."), required=True
+        )
+        type = graphene.Argument(
+            TransactionEventTypeEnum,
+            required=True,
+            description="Current status of the event to report.",
+        )
+        amount = PositiveDecimal(
+            description="The amount of the event to report.", required=True
+        )
+        time = graphene.DateTime(
+            description=(
+                "The time of the event to report. If not provide, "
+                "the current time will be used."
+            )
+        )
+        external_url = graphene.String(
+            description=(
+                "The url that will allow to redirect user to "
+                "payment provider page with event details."
+            )
+        )
+        message = graphene.String(description="The message related to the event.")
+        available_actions = graphene.List(
+            graphene.NonNull(TransactionActionEnum),
+            description="List of all possible actions for the transaction",
+        )
```

Comment:
Usually, when we have so many input fields we create a particular input type for that, what do you think about that?

Reply:
After that the tests could be simplified - the constant could be created for the query.

Reply:
We were thinking with @patrys about having flat input fields here, instead of having `input`. But agree with the `tests` argument.

Reply:
What were the arguments for flat inputs?

Reply:
@IKarbowiak Sorry for the late response. I had some things that I had to handle first. 
Honestly, I don't remember all arguments and prons/cons, but we were generally thinking about using a flat structure instead of injected input type. In that case, the inputs are related to the transactionEvent and TransactionItem (some of them are used to create new transaction event, but some of them will be used to make changes to the transaction item). The main task for the mutation is not creating a new TransactionEvent, but reporting the payment action.